src/lib/tls/tls_signature_scheme.cpp

```diff
@@ -256,6 +258,76 @@ AlgorithmIdentifier Signature_Scheme::algorithm_identifier() const noexcept
+   //    in [SHS]. The length of the Salt MUST be equal to the length of the
+   //    digest algorithm. [...]  When used in certificate signatures, the
+   //    algorithm parameters MUST be DER encoded.
+   auto pss_params = [](const std::string hash_name)
+      {
+         const AlgorithmIdentifier hash_id(hash_name, AlgorithmIdentifier::USE_NULL_PARAM);
+         const AlgorithmIdentifier mgf_id("MGF1", hash_id.BER_encode());
+
+         std::vector<uint8_t> parameters;
+         DER_Encoder(parameters)
+            .start_sequence()
+            .start_context_specific(0).encode(hash_id).end_cons()
+            .start_context_specific(1).encode(mgf_id).end_cons()
+            .start_context_specific(2).encode(HashFunction::create(hash_name)->output_length()).end_cons()
+            .start_context_specific(3).encode(size_t(1)).end_cons() // trailer field
+            .end_cons();
+
+         return parameters;
+      };
+
+   switch(m_code)
+      {
+      // for RSA PKCSv1.5 parameters "SHALL" be NULL
+      case RSA_PKCS1_SHA1:
+         return AlgorithmIdentifier("RSA/EMSA3(SHA-160)", AlgorithmIdentifier::USE_NULL_PARAM);
+      case RSA_PKCS1_SHA256:
+         return AlgorithmIdentifier("RSA/EMSA3(SHA-256)", AlgorithmIdentifier::USE_NULL_PARAM);
+      case RSA_PKCS1_SHA384:
+         return AlgorithmIdentifier("RSA/EMSA3(SHA-384)", AlgorithmIdentifier::USE_NULL_PARAM);
+      case RSA_PKCS1_SHA512:
+         return AlgorithmIdentifier("RSA/EMSA3(SHA-512)", AlgorithmIdentifier::USE_NULL_PARAM);
+
+      // for DSA, ECDSA, GOST parameters "SHALL" be empty
+      case ECDSA_SHA1:
+         return AlgorithmIdentifier("ECDSA/EMSA1(SHA-160)", AlgorithmIdentifier::USE_EMPTY_PARAM);
+      case ECDSA_SHA256:
+         return AlgorithmIdentifier("ECDSA/EMSA1(SHA-256)", AlgorithmIdentifier::USE_EMPTY_PARAM);
+      case ECDSA_SHA384:
+         return AlgorithmIdentifier("ECDSA/EMSA1(SHA-384)", AlgorithmIdentifier::USE_EMPTY_PARAM);
+      case ECDSA_SHA512:
+         return AlgorithmIdentifier("ECDSA/EMSA1(SHA-512)", AlgorithmIdentifier::USE_EMPTY_PARAM);
+
+      case RSA_PSS_SHA256:
+         return AlgorithmIdentifier("RSA/EMSA4", pss_params("SHA-256"));
+      case RSA_PSS_SHA384:
+         return AlgorithmIdentifier("RSA/EMSA4", pss_params("SHA-384"));
+      case RSA_PSS_SHA512:
+         return AlgorithmIdentifier("RSA/EMSA4", pss_params("SHA-512"));
+
+      // for DSA, ECDSA, GOST parameters "SHALL" be empty
+      case DSA_SHA1:
+         return AlgorithmIdentifier("DSA/EMSA1(SHA-160)", AlgorithmIdentifier::USE_EMPTY_PARAM);
+      case DSA_SHA256:
+         return AlgorithmIdentifier("DSA/EMSA1(SHA-256)", AlgorithmIdentifier::USE_EMPTY_PARAM);
+      case DSA_SHA384:
+         return AlgorithmIdentifier("DSA/EMSA1(SHA-384)", AlgorithmIdentifier::USE_EMPTY_PARAM);
+      case DSA_SHA512:
+         return AlgorithmIdentifier("DSA/EMSA1(SHA-512)", AlgorithmIdentifier::USE_EMPTY_PARAM);
```

Comment:
Hold on a minute - do we still support DSA ciphersuites?

It appears we do?? Or at least there is something left over after #2505 which I had though removed DSA entirely...

A 3.0 release seems like a very opportune time to drop all support for DSA from TLS, if we haven't already ...

Reply:
OK it appears the only real usage is in the `Signature_Scheme::Code` - will open a PR

Reply:
Removed in #3143